src/compiler/crystal/tools/print_hierarchy.cr

```diff
@@ -260,4 +267,107 @@ module Crystal
       ::with_color.toggle(@program.color?)
     end
   end
+
+  class JSONHierarchyPrinter < HierarchyPrinter
+    def execute
+      if exp = @exp
+        compute_targets(@program.types, exp, false)
+      end
+
+      json = String.build do |io|
```

Comment:
From here, to the end of the method, this should be:

```cr
      STDOUT.json_object do |json_object|
        print_type(@program.object, json_object, STDOUT)
      end
```

otherwise you are streaming to a string in memory, thus creating an object, which is what we wanted to avoid in the first place.

That won't compile because of the type restrictions bellow (`JSON::ObjectBuilder(String::Builder)`). Just remove those restrictions, they aren't necessary here. Then it will compile.